The great toxification (part 5)

stephenfin opened the pull request:
Subject: Finish things off in preparation for switching CIs to `tox`.

We want to switch CIs to use `tox` instead of the Makefile. Before we can do so, we need to close out a few more small things. In this series, we make the following changes:

- Take advantage of pytest features to dynamically disable tests if:
  - Dependencies are missing (e.g. `dvipng` or `dvisvgm`)
  - The version of Python is not high enough (e.g. only Python > 3.6 supports the `async` keyword)
- Take advantage of `tox` features to massively reduce the amount of duplication in the tox file

This is best reviewed by looking at commits individually [here](https://github.com/sphinx-doc/sphinx/pull/4175/commits). I use commits to break this up into readable chunks so it can be reviewed.

### Feature or Bugfix
- Feature

### Purpose
To paraphrase the Zen of Python (import this):

> tox is one honking great idea -- let's use more of it!

Start modernizing the test infrastructure that's using make and other hand-crafted tools by replacing them with tox-based derivatives.

**This is part five, step one.**

### Detail
This is a multi-step process, with the end goal of completely removing the Makefile and running everything through `tox`. The first fours steps of these can be done in parallel, but combined they block the final two.

1. Make `tox` a little more friendly by enabling extras like coverage, test profiling etc. by default, and closing gaps with the current Makefile targets such as PyLint and building docs in particular formats
2. Replace/remove custom tooling found in `utils` in favor of modern alternatives like `flake8`
3. Centralize as much configuration as possible inside `setup.cfg` and `tox.ini`, to avoid polluting the top level directory.
4. Clean up requirements (`test-reqs.txt`, `setup.py`) to ensure virtualenv and non-virtualenv builds are identical

The remaining two steps are blocked by the above:

5. Switch CIs to use `tox` instead of the Makefile, and update `CONTRIBUTING.rst` to refer to tox. Deprecate the Makefile  (**this change**)
6. (Bonus) Integrate [codecov.io](https://codecov.io) to make use of the coverage stats we now have

### Relates
- N/A

stephenfin committed:
tests: Skip tests with missing binaries

While there are already some skips included here, they clearly aren't
doing their job and the test fail locally. Resolve this.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
setup.py: Include 'flake8' in 'test' requirements

We are using this in our testing and actually need it for the flake8
plugin, so include this in the list of 'test' requirements.

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented:
It seems failed. Could you check again?
In addition, execution time of testing is much increased. Please check that also.

stephenfin commented:
> It seems failed. Could you check again?

Fixed. It was a dependency issue.

> In addition, execution time of testing is much increased. Please check that also.

This is increased because I'm now testing far more different environments (e.g. running different versions of docutils against all versions of Python). I've reduced a little of the duplication now though so it should be almost the same running time

tk0miya commented:
Okay, please let me know after all your work finished. I will review this again.

stephenfin commented:
@tk0miya This is resolved now. Run times are still slightly above where they were, but that's because I've enabled coverage for all targets except PyPy (see commits for rationale). I need that data to do step 6. Let me know if this is acceptable for you.

tk0miya commented:
Now, this PR takes about 1hour for testing.
>Ran for 14 min 40 sec
>Total time 1 hr 1 min 47 sec

But, current master branch takes about 30+min for testing.
>Ran for 11 min 28 sec
>Total time 39 min 25 sec
>https://travis-ci.org/sphinx-doc/sphinx/builds/291345854

>Ran for 9 min 22 sec
>Total time 32 min 37 sec
>https://travis-ci.org/sphinx-doc/sphinx/builds/291043182

>Ran for 8 min 26 sec
>Total time 33 min 20 sec
>https://travis-ci.org/sphinx-doc/sphinx/builds/290779390

I have two question for this PR.

* Is it important to measure coverage with all interpreters?
* Is `tox-travis` useful for our project? I feel it reduces concurrency.

tk0miya commented:
(Now I restart Travis process manually to measure a time again)

As a result, it also takes 1 hour.
>Ran for 15 min 50 sec
>Total time 57 min 13 sec

Note: new flake8 package is released just now. And it causes errors. I'll fix it soon.

stephenfin commented:
> I have two question for this PR.
> 
> Is it important to measure coverage with all interpreters?

I think it's important to measure coverage with a Python 2.7 and Python 3.x interpreter. However, as we do not currently generate coverage figures at all, I can reduce this to Python 3.x.

> Is tox-travis useful for our project? I feel it reduces concurrency.

The idea of this is to make sure we're testing in the gate the same way that we are testing locally. If I can get the test time the same then I think it's a good move.

stephenfin committed:
tests: Ignore tests using 'collect_ignore'

Per the pytest docs [1], this is the preferred way to ignore tests. This
necessitates removing the 'test-async' target as it no longer makes any
sense.

[1] https://docs.pytest.org/en/latest/example/pythoncollection.html

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
tests: Ignore roots using 'collect_ignore'

This is slightly cleaner than how we're doing this at the moment.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
tox: Add 'coverage' target

Enabling coverage results in an increase in runtime, and while this may
be an acceptable compromise for some developers, it's not necessarily
something we need to run all the time. Move it to a separate target so
it can be run only by those who wish to use it.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin commented:
@tk0miya I've switched coverage to only run against one version and greatly simplified the tox file. Let's see how the times are this go round.

stephenfin committed:
tox: Make further use of factor-conditional deps

This is how we reduce boilerplate [1]. Succinctness FTW.

[1] https://tox.readthedocs.io/en/latest/config.html#factors-and-factor-conditional-settings

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented on .travis.yml:
```diff
@@ -2,46 +2,43 @@ language: python
   - "3.5"
   - "3.4"
   - "2.7"
-  - "nightly"
```
Why removing `nightly` here?

stephenfin replied:
My mistake. Re-added now.

tk0miya commented on CONTRIBUTING.rst:
```diff
@@ -33,7 +33,7 @@ Bug Reports and Feature Requests
 If you have encountered a problem with Sphinx or have an idea for a new
 feature, please submit it to the `issue tracker`_ on GitHub or discuss it
-on the sphinx-dev mailing list.
+on the `sphinx-dev` mailing list.
```
Please use double back quotes for literals instead.
http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#inline-literals

stephenfin replied:
I don't want to make this literal - I want to make it italic. There are other occurrences where I should be using literals though so I've fixed those

tk0miya replied:
okay, please use emphasis notation.
http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#emphasis

In reST, single back quote is used for "interpreted text". So it is recognized as a kind of role.
http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#interpreted-text

tk0miya commented on CONTRIBUTING.rst:
```diff
@@ -112,39 +116,48 @@ These are the basic steps needed to start developing on Sphinx.
+
+         tox -v
+
+   * To run unit tests for a specific Python version, such as 2.7::
```
I prefer to use 3.6 for this example. 2.7 is too old...

stephenfin replied:
Done

tk0miya reviewed (changes requested):
LGTM with nits.

About running time, I'll measure it in some times from now on.

tk0miya commented:
Note:

1st time:
```
 Ran for 12 min 9 sec
 Total time 43 min 21 sec
```

2nd time:
```
 Ran for 4 min 21 sec
 Total time 36 min 31 sec
```
(But pypy takes `9 min 40 sec`)

3rd time:
```
 Ran for 34 min 56 sec
 Total time 1 hr 51 min 47 sec
```
(wow! I guess some heavy process was invoked in same VM.)

4th time:
```
 Ran for 32 min 1 sec
 Total time 1 hr 37 min 22 sec
```

stephenfin commented:
@tk0miya I have fixed the issues at hand. Latest build time was "44 min 44 sec", which seems correct given that we're installing many more dependencies. I also removed the use of `tox-travis` as things were getting slightly confusing.

Is there anything else needed here?

stephenfin commented:
@tk0miya Is there anything else I need here?

stephenfin commented:
@tk0miya I've taken out the steps to enable `tox` in the CI to make this even smaller. Hopefully this is OK for you to review.

stephenfin commented on Makefile:
```diff
@@ -69,11 +69,11 @@ reindent:
 
 .PHONY: test-async
 test-async:
-	@cd tests; $(PYTHON) run.py -v $(TEST)
```
This is no longer necessary because of the changes I made in `tests/conftest.py`.

tk0miya commented:
Sorry for very late reaction. LGTM!

stephenfin commented:
> Sorry for very late reaction. LGTM!

No problem. Thanks for the merge 👍
